overlord/snapstate/snapstate_update_test.go

```diff
@@ -9925,3 +9925,148 @@ func (s *snapmgrTestSuite) TestMonitoringIsPersistedAndRestored(c *C) {
+		case "snapd":
+			if si.Revision.N == 1 {
+				version = old
+			} else if si.Revision.N == 2 {
+				version = new
+			} else {
+				return nil, fmt.Errorf("unexpected revision for test")
+			}
+		default:
+			version = "1.0"
+		}
+		return &snap.Info{
+			SuggestedName: name,
+			Version:       version,
+			Architectures: []string{"all"},
+			SideInfo:      *si,
+		}, nil
+	})
+	defer restore()
+
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("unrelated", "...")
+	chg.AddTask(st.NewTask("task0", "..."))
+
+	snapstate.Set(s.state, "snapd", &snapstate.SnapState{
+		Active:          true,
+		Sequence:        []*snap.SideInfo{&si14, &si5},
+		TrackingChannel: "latest/stable",
+		Current:         si14.Revision,
+	})
+
+	_, err := snapstate.Update(s.state, "snapd", &snapstate.RevisionOptions{Revision: snap.R(2)}, s.user.ID, snapstate.Flags{})
+	return err
+}
+
+func (s *snapmgrTestSuite) TestUpdateDowngradeBlockedByOtherChanges(c *C) {
+	err := s.testUpdateDowngradeBlockedByOtherChanges(c, "2.57.1", "2.56")
+	c.Assert(err, ErrorMatches, `other changes in progress \(conflicting change "unrelated"\), change "snapd-downgrade" not allowed until they are done`)
+}
+
+func (s *snapmgrTestSuite) TestUpdateDowngradeNotBlockedByOtherChanges(c *C) {
+	err := s.testUpdateDowngradeBlockedByOtherChanges(c, "2.57.1", "2.58")
+	c.Assert(err, IsNil)
+}
+
+func (s *snapmgrTestSuite) testUpdateNotAllowedWhileDowngrading(c *C, old, new string) error {
+	si14 := snap.SideInfo{
+		RealName: "snapd",
+		SnapID:   "snapd-id",
+		Channel:  "latest",
+		Revision: snap.R(1),
+	}
+	si5 := snap.SideInfo{
+		RealName: "snapd",
+		SnapID:   "snapd-id",
+		Channel:  "latest",
+		Revision: snap.R(2),
+	}
+
+	si := snap.SideInfo{
+		RealName: "some-snap",
+		SnapID:   "some-snap-id",
+		Revision: snap.R(7),
+		Channel:  "channel-for-7",
+	}
+
+	restore := snapstate.MockSnapReadInfo(func(name string, si *snap.SideInfo) (*snap.Info, error) {
+		var version string
+		switch name {
+		case "snapd":
+			if si.Revision.N == 1 {
+				version = old
+			} else if si.Revision.N == 2 {
+				version = new
+			} else {
+				return nil, fmt.Errorf("unexpected revision for test")
+			}
+		default:
+			version = "1.0"
+		}
+		return &snap.Info{
+			SuggestedName: name,
+			Version:       version,
+			Architectures: []string{"all"},
+			SideInfo:      *si,
+		}, nil
+	})
+	defer restore()
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	snapstate.Set(s.state, "snapd", &snapstate.SnapState{
+		Active:          true,
+		Sequence:        []*snap.SideInfo{&si14, &si5},
+		TrackingChannel: "latest/stable",
+		Current:         si14.Revision,
+	})
+	snapstate.Set(s.state, "some-snap", &snapstate.SnapState{
+		Active:          true,
+		Sequence:        []*snap.SideInfo{&si},
+		TrackingChannel: "other-chanel/stable",
+		Current:         si.Revision,
+	})
+
+	ts, err := snapstate.Update(s.state, "snapd", &snapstate.RevisionOptions{Revision: snap.R(2)}, s.user.ID, snapstate.Flags{})
```

Comment:
we also need a similar test with revert?

Reply:
Really nice catch! Added revert unit test cases!